assembly/frame/system.ts

```diff
@@ -0,0 +1,232 @@
+    // nonce key
+    static readonly NONCE_KEY: string = "nonce";
+    /**
+     * System storage keys
+     */
+    // execution phase
+    static readonly EXEC_PHASE: string[] = ["system", "exec_phase"];
+    // parent hash
+    static readonly PARENT_HSH: string[] = ["system", "parent_hsh"];
+    // block number
+    static readonly BLOCK_NUM0: string[] = ["system", "block_num0"];
+    // extrinsics root
+    static readonly EXTCS_ROOT: string[] = ["system", "extcs_root"];
+    // digest items
+    static readonly DIGESTS_00: string[] = ["system", "digests_00"];
+    // block hash
+    static readonly BLOCK_HASH: string[] = ["system", "block_hash"];
+    // extrinsics count
+    static readonly EXTCS_COUT: string[] = ["system", "extcs_cout"];
+    // extrinsics data 
+    static readonly EXTCS_DATA: string[] = ["system", "extcs_data"];
+    // block hash count (max number of blocks to be stored)
+    static readonly BHSH_COUNT: string[] = ["system", "bhash_cout"];
+    
+    /**
+     * @description Sets up the environment necessary for block production
+     * @param header Header instance
+    */
+   static initialize(header: HeaderType): void{
+    // maximum number of blocks
+    const bhshCount = RuntimeConstants.blockHashCount();
+    Storage.set(Utils.stringsToBytes(this.BHSH_COUNT, true), bhshCount.toU8a());
+    Storage.set(Utils.stringsToBytes([this.EXTRINSIC_INDEX], true), [<u8>0]);
+    Storage.set(Utils.stringsToBytes(this.EXEC_PHASE, true), Utils.stringsToBytes([System.INITIALIZATION], true));
+    Storage.set(Utils.stringsToBytes(this.PARENT_HSH, true), header.getParentHash().toU8a());
+    Storage.set(Utils.stringsToBytes(this.BLOCK_NUM0, true), header.getNumber().toU8a());
+    Storage.set(Utils.stringsToBytes(this.EXTCS_ROOT, true), header.getExtrinsicsRoot().toU8a());
+
+    let digests: u8[] = [];
+    for(let i: i32 = 0; i < header.getDigests().length; i++){
+        digests.concat(header.getDigests()[i].toU8a());
+    }
+
+    Storage.set(Utils.stringsToBytes(this.DIGESTS_00, true), digests);
+    const blockNumber: BlockNumber = instantiate<BlockNumber>(header.getNumber().value - 1);
+    this.setHashAtBlock(blockNumber, <HashType>header.getParentHash());
+}
+    /**
+     * @description Removes temporary "environment" entries in storage and finalize block
+     */
+    static finalize(): HeaderType {
+        Storage.clear(Utils.stringsToBytes(this.EXEC_PHASE, true));
+        Storage.clear(Utils.stringsToBytes(this.EXTCS_COUT, true));
+        let blockNumber = Storage.take(Utils.stringsToBytes(this.BLOCK_NUM0, true));
+        let parentHash = Storage.take(Utils.stringsToBytes(this.PARENT_HSH, true));
+        let digests = Storage.take(Utils.stringsToBytes(this.DIGESTS_00, true));
+        let extrinsicsRoot = Storage.take(Utils.stringsToBytes(this.EXTCS_ROOT, true));
+        
+        // move block hash pruning window by one block
+        let blockHashCount = <u32>(this.blockHashCount().value);
```

Comment:
I am not sure that we should cast to `u32` both of them. What if block number is u64?

I think that `blockHashCount` should use the same user-defined runtime type as `blockNumber` this way we will be able to do `blockNumber.value > blockHashCount.value`